src/LanguageServerProtocol.fs

```diff
@@ -2707,70 +2708,60 @@ type LspServer() =
+    let requestHandling<'param, 'result, 'server when 'server :> LspServer> (run: 'server -> 'param -> AsyncLspResult<'result>): RequestHandling<'server> =
+        let tokenRun (decoration: RequestHandlingDecoration) (lspServer: 'server) : Delegate =
+            let runAndUnwrap param = async {
+                match decoration with
+                | Some fn -> fn()
+                | None -> ()
+
+                let! lspResult = run lspServer param
 
-    let serialize<'t> (o: 't) = JToken.FromObject(o, jsonSerializer)
+                return
+                    match lspResult with
+                    | Ok result -> result
+                    | Error error ->
+                        let rpcException = LocalRpcException(error.Message)
+                        rpcException.ErrorCode <- error.Code
+                        rpcException.ErrorData <- error.Data |> Option.defaultValue null
+                        raise rpcException
+            }
 
-    let requestHandling<'param, 'result, 'server when 'server :> LspServer> (run: 'server -> 'param -> AsyncLspResult<'result>): RequestHandling<'server> =
-        let tokenRun (lspServer: 'server) (paramsToken: JToken option): AsyncLspResult<JToken option> =
-            try
-                let p =
-                    match paramsToken with
-                    | Some paramsToken ->
-                        let typedParams = paramsToken.ToObject<'param>(jsonSerializer)
-                        Some typedParams
-                    | None ->
-                        if typeof<'param> = typeof<unit> then
-                            Some (unbox ())
-                        else
-                            None
-                match p with
-                | Some p ->
-                    async {
-                        try
-                            let! result = run lspServer p
-                            match result with
-                            | Result.Ok ok ->
-                                if isNull (box ok) then
-                                    return Result.Ok None
-                                else
-                                    let resultToken = JToken.FromObject(ok, jsonSerializer)
-                                    return Result.Ok (Some resultToken)
-                            | Result.Error err ->
-                                return Result.Error err
-                        with
-                        | ex ->
-                            return Result.Error (Error.Create(ErrorCodes.internalError, ex.ToString()))
-                    }
-                | None ->
-                     async.Return (Result.Error (Error.Create(ErrorCodes.invalidRequest, "No params found")))
-            with
-            | :? JsonException as ex ->
-                async.Return (Result.Error (Error.Create(ErrorCodes.parseError, ex.ToString())))
+            let runAsTask param ct =
+                Async.StartAsTask(runAndUnwrap param, cancellationToken=ct)
```

Comment:
@baronfel I am in a bit of a pickle here with request ordering

Basically, StreamJsonRpc provides good request ordering, but it does that in a TPL/C# Task-specific way.. it ensures another task delegate is not invoked until the current one returns a Task so the application can enqueue current request before allowing the next one to come in. See first paragraph in https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/resiliency.md#nonconcurrentsynchronizationcontext

For C# it is relatively ok, as we could just do all the queuing of the work in non-async code before returning a Task from the delegate method. But in F# I am forced to launch Async.StartAsTask, if I want to attach a cancellation token to async workflow... And F# async are "cold" tasks.. So I was mucking with this code a bit and I probably don't know how to accomplish that thing with current prototype for `(run: 'server -> 'param -> AsyncLspResult<'result>)`.

Maybe you have an idea. Otherwise I will have to change the prototype of `run` to return a `Task<LspResult>` instead to account for StreamJsonRpc sequencing signaling in my handler code that does enqueue text changes from the client to roslyn model in good sequence before returning a `task` that can then be run on task pool as it is ok at that point for me.

Is there some way to set Async.CancellationToken w/o invoking Async.StartAsTask.. I am not sure myself what would be the correct approach before converting the thing to use `task`s..

This is mostly evident in csharp-ls when there are multiple `textDocument/didChange` requests from the client and those are dumped to TPL task pool but are executed randomly(?) The result is I process document deltas in wrong order the document model gets corruped.

Reply:
It's probably ok to just accept needing to use tasks. That's what I did with my prior attempt, and we have good interop with that now in f# 6/with ply et al.

Reply:
Hmm, then I will have to pass in ct to every request handler, that patch will get ugly :/

Reply:
Yeah I know. It was a lot of churn in FSAC when I attempted it in the past, but I think it's worth it to take the pain. We get a well supported and fast json rpc base, and in exchange we have to be a bit more explicit about things.

Reply:
hmm, I actually managed to avoid the need to convert everything to task-based flow. I simply do it like for tasks -- return async computation/object/block (not sure what is the right name for that ?) at the last moment and thus I have chance to obtain state locks and to nott drop to TPL thread pool until I have things in place when handling a request.

I will need to update this PR later.

Also, to make it easier for me to compose handler behaviour I will introduce another version of `start` function that does not require a server "class" -- just a map of handlers. I loose some of type checking (I guess..) but I can compose the behaviour more easily with functional techniques. I will leave the original version in place too, so we should be good.